libstratum/EthStratumClient.cpp

```diff
@@ -532,16 +532,16 @@ bool EthStratumClient::submit(Solution solution) {
 		minernonce = nonceHex.substr(m_extraNonceHexSize, 16 - m_extraNonceHexSize);
 	}
 
-	if (EthashAux::eval(tempWork.seed, tempWork.header, solution.nonce).value < tempWork.boundary)
+	if (!solution.stale)
 	{
 		string json;
 
 		switch (m_protocol) {
 			case STRATUM_PROTOCOL_STRATUM:
-				json = "{\"id\": 4, \"method\": \"mining.submit\", \"params\": [\"" + p_active->user + "\",\"" + temp_job + "\",\"0x" + nonceHex + "\",\"0x" + tempWork.header.hex() + "\",\"0x" + solution.mixHash.hex() + "\"]}\n";
+				json = "{\"id\": 4, \"method\": \"mining.submit\", \"params\": [\"" + p_active->user + "\",\"" + temp_job + "\",\"0x" + nonceHex + "\",\"0x" + temp_header.hex() + "\",\"0x" + solution.mixHash.hex() + "\"]}\n";
```

Comment:
Why didn't you use the header hash from `solution`? `solution.headerHash`.

Reply:
Darn! I hadn't seen it. Good catch. Too bad the job_name isn't in there as well... with the header hash already in the solution, we only need to keep a copy of the previous job_name, not the whole work package. I'll try it out.

Reply:
Still, wouldn't need any of this if we just threw again known stale shares! Do you know why stales are even sent?

Reply:
i'm not 100% sure but i think some pools still pay stale shares. Even if not, they provide graphical stats of stale shares

Reply:
Yes, they should pay for it.

If you need the job id, add it to WorkPackage and Solution. Trying to match solution with job id by stratum thread is bad idea.